src/index/Vocabulary.cpp

```diff
@@ -267,8 +327,74 @@ auto Vocabulary<S, C, I>::prefixRanges(std::string_view prefix) const
 template <typename S, typename C, typename I>
 auto Vocabulary<S, C, I>::operator[](IndexType idx) const
     -> AccessReturnType_t<S> {
-  AD_CONTRACT_CHECK(idx.get() < size());
-  return vocabulary_[idx.get()];
+  // Check marker bit to determine which vocabulary to use
+  if (idx.get() & geoVocabMarker) {
+    // The requested word is stored in the geometry vocabulary
+    uint64_t unmarkedIdx = idx.get() & geoVocabIndexMask;
+    return geoVocabulary_[unmarkedIdx];
+  } else {
+    // The requested word is stored in the vocabulary for normal words
+    AD_CONTRACT_CHECK(idx.get() <= maxWordIndex);
+    return vocabulary_[idx.get()];
+  }
+}
+
+// _____________________________________________________________________________
+template <class S, class C, typename I>
+Vocabulary<S, C, I>::WordWriter::WordWriter(
+    const VocabularyWithUnicodeComparator& vocabulary,
+    const std::string& filename)
+    : underlyingWordWriter_{vocabulary.getUnderlyingVocabulary().makeDiskWriter(
+          filename)},
+      underlyingGeoWordWriter_{
+          vocabulary.getUnderlyingVocabulary().makeDiskWriter(
+              filename + geoVocabSuffix)} {};
+
+// _____________________________________________________________________________
+template <class S, class C, typename I>
+uint64_t Vocabulary<S, C, I>::WordWriter::operator()(std::string_view word,
+                                                     bool isExternal) {
+  if (stringIsGeoLiteral(word)) {
+    // The word to be stored in the vocabulary is a geometry literal. It
+    // needs to be written to the dedicated geometry vocabulary and get an
+    // index with the marker bit set to 1.
+    uint64_t index;
+    if constexpr (std::is_same_v<S, CompressedString>) {
```

Comment:
Comment for later : as soon as we have merged the other PR, this if can probably go away.